includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php

```diff
@@ -1220,6 +1227,16 @@ public function get_pages() {
 				}
 			}
 
+			// Block checks.
+			if ( $values['block'] && get_post( $page_id ) ) {
+				$block_required = true;
+				$page           = get_post( $page_id );
+				$page_info      = WC_Tracker::get_block_tracker_data( $values['block'], $page_name );
```

Comment:
I feel a bit weird calling `WC_Tracker` for something which is not really related to the tracker.

Out of the scope for this PR, but at some point we might want to abstract that method somewhere outside the tracker.

Reply:
Absolutely, it feels wired, but the use is so small that I did not want to abstract too early.

Reply:
FYI the checks in `get_block_tracker_data` are a bit primitive, could be improved - https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/3224#issuecomment-707391821

There's a new improved approach using `parse_blocks()` in a PR that didn't merge, could be adapted into core: https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/bbef0a6763f3c5073902be248820badc09bc2496

It's might be worthwhile spending a bit of time tidying this up and abstracting out the detection logic. We don't often touch this code, so possibly worth investing a little more effort?

Reply:
I don't mind doing it, cleanup makes sense. But I don't want to do it as a part of this PR. The scope should be separated. I don't mind postponing this and working on a cleaned-up version if this is what we think this is how this should be done.

Reply:
Or merge as it is and follow up with a cleanup.

Reply:
A couple thoughts from my perspective here:

- I don't like using `WC_Tracker:: get_block_tracker_data` as this on the surface appears to be a violation of separation of concerns. For now the method does what you want, but it is a part of a class dedicated to WC_Tracker events so there's no guarantee it will do what you want in future refactors (or do what you want without any side-effects).
- To that point, extraction of whether blocks are in a page should not even be a method on `WC_Tracker` itself but rather something `WC_Tracker` consumes as a service.

I recognize the desire to keep scope from creeping in this pull and normally I'd agree, but this is severe enough that I'd like to avoid compounding the tech debt that already exists by pulling in `WC_Tracker` usage. I took a brief look at the code Rua shared and it should be trivial enough to create a new service class in Woo for explicitly solving this block detection problem (using the code Rua shared, or some variation of it). If you want to avoid creating an entire new class, you can add it as private methods in this class (`WC_REST_System_Status_V2_Controller`) for now and future follow ups could extract to a service class and implement in WC_Tracker too. I'd prefer creating the class.